back/boxtribute_server/business_logic/statistics/crud.py

```diff
@@ -206,3 +215,55 @@ def compute_top_products_donated(base_id):
 
     dimensions = _generate_dimensions("category", "product", "size", facts=facts)
     return {"facts": facts, "dimensions": dimensions}
+
+
+TARGET_ID_SEPARATOR = "---"
+
+
+def compute_moved_boxes(base_id):
+    """Count all boxes moved to locations in the given base, grouped by date of
+    movement, product category, and box state.
+    """
+    # This selects only information of boxes that were moved from InStock to Donated
+    # state, and are now in the base of given base ID. It is NOT taken into account that
+    # boxes can be moved back from Donated to InStock, nor that the product or other
+    # attributes of the box change after having been donated
+    selection = (
+        DbChangeHistory.select(
+            fn.MAX(DbChangeHistory.change_date).alias("moved_on"),
+            fn.CONCAT(Base.name, TARGET_ID_SEPARATOR, Location.name).alias("target_id"),
+            Product.category.alias("category_id"),
+            fn.COUNT(Box.id).alias("boxes_count"),
+        )
+        .join(
+            Box,
+            on=(
+                (DbChangeHistory.record_id == Box.id)
+                & (DbChangeHistory.table_name == Box._meta.table_name)
+                & (DbChangeHistory.changes == Box.state.column_name)
+                & (DbChangeHistory.from_int == BoxState.InStock)
+                & (DbChangeHistory.to_int == BoxState.Donated)
+            ),
+        )
+        .join(
+            Product,
+            on=((Box.product == Product.id) & (Product.base == base_id)),
+        )
+        .join(
+            Location,
+            src=Box,
+            on=((Box.location == Location.id) & (Location.base == base_id)),
+        )
+        .join(Base)
+    )
+    facts = selection.group_by(
+        SQL("target_id"),
+        SQL("category_id"),
```

Comment:
@pylipp Shouldn't you also group by "movedOn"? I'm not really familiar in peewee, but in SQL I would have to do that.

Reply:
When I do that the query errors with `1056 - Can't group on 'moved_on'`. I think this is because the `moved_on` column is derivate by the MAX aggregate.

Reply:
ahhh, yes @pylipp . was already wondering cause the results were also off.

So, you have to do two sequential group bys - one group by in a subquery over record_id in history to get the movedOn date and then one group by over movedOn, target_id and category_id.

Reply:
@pylipp almost there, but not completely. 
`moved_on` in your query is the date of the last entry of the corresponding `record_id` in history.
What you need is the last date of the corresponding `record_id` WHERE the tablename is stock, the change is Box_state_id, from_to=......

Therefore, you will need to move this ON clause  `AND (`t1`.`tablename` = %s)) AND (`t1`.`changes` = %s)) AND (`t1`.`from_int` = %s)) AND (`t1`.`to_int` = %s))`
to a WHERE clause in the subquery sq.